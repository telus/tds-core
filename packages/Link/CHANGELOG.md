# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.2.6](https://github.com/telusdigital/tds/compare/@tds/core-link@2.2.5...@tds/core-link@2.2.6) (2020-07-22)

**Note:** Version bump only for package @tds/core-link





## [2.2.5](https://github.com/telusdigital/tds/compare/@tds/core-link@2.2.4...@tds/core-link@2.2.5) (2020-03-25)


### Bug Fixes

* **core-link:** prop type warnings for reactRouterLinkComponent ([fcf1189](https://github.com/telusdigital/tds/commit/fcf1189827ad37ee3429e970019c87e6783f40b0)), closes [#1401](https://github.com/telusdigital/tds/issues/1401)





## [2.2.4](https://github.com/telusdigital/tds/compare/@tds/core-link@2.2.3...@tds/core-link@2.2.4) (2020-03-17)

**Note:** Version bump only for package @tds/core-link





## [2.2.3](https://github.com/telusdigital/tds/compare/@tds/core-link@2.2.2...@tds/core-link@2.2.3) (2020-02-03)

**Note:** Version bump only for package @tds/core-link





## [2.2.2](https://github.com/telusdigital/tds/compare/@tds/core-link@2.2.1...@tds/core-link@2.2.2) (2020-01-27)

**Note:** Version bump only for package @tds/core-link





## [2.2.1](https://github.com/telusdigital/tds/compare/@tds/core-link@2.2.0...@tds/core-link@2.2.1) (2020-01-23)

**Note:** Version bump only for package @tds/core-link





# [2.2.0](https://github.com/telusdigital/tds/compare/@tds/core-link@2.1.3...@tds/core-link@2.2.0) (2020-01-22)


### Features

* **core-link:** add icon ([f4aca15](https://github.com/telusdigital/tds/commit/f4aca15))





## [2.1.3](https://github.com/telusdigital/tds/compare/@tds/core-link@2.1.2...@tds/core-link@2.1.3) (2020-01-07)


### Bug Fixes

* **core-link:** default invert prop to undefined to prevent router prop type error ([ef84f42](https://github.com/telusdigital/tds/commit/ef84f42))





## [2.1.2](https://github.com/telusdigital/tds/compare/@tds/core-link@2.1.1...@tds/core-link@2.1.2) (2019-10-09)

**Note:** Version bump only for package @tds/core-link





## [2.1.1](https://github.com/telusdigital/tds/compare/@tds/core-link@2.1.0...@tds/core-link@2.1.1) (2019-10-04)

**Note:** Version bump only for package @tds/core-link





# [2.1.0](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.11...@tds/core-link@2.1.0) (2019-09-18)


### Features

* **core-link:** allow forwarding refs ([d992dd8](https://github.com/telusdigital/tds/commit/d992dd8))





## [2.0.11](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.10...@tds/core-link@2.0.11) (2019-09-12)

**Note:** Version bump only for package @tds/core-link





## [2.0.10](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.9...@tds/core-link@2.0.10) (2019-09-06)

**Note:** Version bump only for package @tds/core-link





## [2.0.9](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.8...@tds/core-link@2.0.9) (2019-08-22)

**Note:** Version bump only for package @tds/core-link





## [2.0.8](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.7...@tds/core-link@2.0.8) (2019-08-09)

**Note:** Version bump only for package @tds/core-link





## [2.0.7](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.6...@tds/core-link@2.0.7) (2019-08-01)

**Note:** Version bump only for package @tds/core-link





## [2.0.6](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.5...@tds/core-link@2.0.6) (2019-07-30)

**Note:** Version bump only for package @tds/core-link





## [2.0.5](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.4...@tds/core-link@2.0.5) (2019-06-25)

**Note:** Version bump only for package @tds/core-link





## [2.0.4](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.3...@tds/core-link@2.0.4) (2019-06-04)

**Note:** Version bump only for package @tds/core-link

## [2.0.3](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.2...@tds/core-link@2.0.3) (2019-05-07)

**Note:** Version bump only for package @tds/core-link

## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.1...@tds/core-link@2.0.2) (2019-05-01)

**Note:** Version bump only for package @tds/core-link

## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-link@2.0.0...@tds/core-link@2.0.1) (2019-05-01)

**Note:** Version bump only for package @tds/core-link

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.6...@tds/core-link@2.0.0) (2019-04-17)

### Features

- **core-link:** upgrade to styled components ([587e220](https://github.com/telusdigital/tds/commit/587e220))

### BREAKING CHANGES

- **core-link:** add React 16.8 and styled-components 4.1.3 to peerDependencies

## [1.0.6](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.5...@tds/core-link@1.0.6) (2019-03-18)

**Note:** Version bump only for package @tds/core-link

## [1.0.5](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.4...@tds/core-link@1.0.5) (2019-03-05)

**Note:** Version bump only for package @tds/core-link

## [1.0.4](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.3...@tds/core-link@1.0.4) (2019-02-06)

**Note:** Version bump only for package @tds/core-link

<a name="1.0.3"></a>

## [1.0.3](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.2...@tds/core-link@1.0.3) (2018-08-30)

### Bug Fixes

- **core-link:** clean up css ([afaaa55](https://github.com/telusdigital/tds/commit/afaaa55))
- **core-link:** prevent link underline from hiding when zooming out ([48b5647](https://github.com/telusdigital/tds/commit/48b5647))
- **core-link:** use text-decoration: underline ([a045d83](https://github.com/telusdigital/tds/commit/a045d83))

<a name="1.0.2"></a>

## [1.0.2](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.1...@tds/core-link@1.0.2) (2018-04-26)

### Bug Fixes

- **core-link:** update link colour to be AAA WCAG compliant ([#601](https://github.com/telusdigital/tds/issues/601)) ([bffc242](https://github.com/telusdigital/tds/commit/bffc242))

<a name="1.0.1"></a>

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-link@1.0.0...@tds/core-link@1.0.1) (2018-03-22)

### Bug Fixes

- **core-link:** standardize focus outline across browsers ([7ba5d17](https://github.com/telusdigital/tds/commit/7ba5d17)), closes [#445](https://github.com/telusdigital/tds/issues/445)
