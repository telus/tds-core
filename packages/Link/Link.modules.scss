@import '../colour/colours';

@mixin add-skinny-underline {
  background: linear-gradient(transparent, transparent), linear-gradient(transparent, transparent),
    linear-gradient(currentColor, currentColor);
  background-size: 0.05em 1px, 0.05em 1px, 1px 1px;
  background-repeat: no-repeat, no-repeat, repeat-x;
  background-position: 0 100%, 100% 100%, 0 100%;
  text-decoration: none;
}

@mixin clear-skinny-underline {
  background-size: 0.05em 1px, 0.05em 1px, 1px 0;
}

.underlineOnHover {
  &:link,
  &:visited {
    @include add-skinny-underline;
  }

  &:hover {
    @include clear-skinny-underline;
  }
}

.base {
  composes: underlineOnHover;

  &:link,
  &:visited {
    color: $color-link;
  }
}

.inverted {
  composes: underlineOnHover;

  &:link,
  &:visited {
    color: $color-white;
  }
}

.inheritColor {
  composes: underlineOnHover;

  &:link,
  &:visited {
    color: inherit;
  }
}
