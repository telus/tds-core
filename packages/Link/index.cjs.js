const Link = require('./dist/index.cjs')

module.exports = Link
