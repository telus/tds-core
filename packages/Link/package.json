{
  "name": "@tds/core-link",
  "version": "2.2.6",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telusdigital/tds.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telusdigital/tds/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/shared-hocs": "^1.2.0",
    "@tds/shared-styles": "^1.5.2",
    "@tds/util-helpers": "^1.4.2",
    "@tds/util-prop-types": "^1.4.0",
    "prop-types": "^15.5.10"
  },
  "devDependencies": {
    "@tds/core-colours": "^2.2.1"
  }
}
