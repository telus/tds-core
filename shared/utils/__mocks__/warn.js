export const deprecate = jest.fn()
export const warn = jest.fn()
