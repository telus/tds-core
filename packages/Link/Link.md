### Minimal usage

```jsx
<Link href="#">Go to the TDS homepage</Link>
```

### Usage criteria

- Links inherit the font size of surrounding text.
- Their colour is either shuttle grey or white.

### Inverted

```jsx { "props": { "className": "docs_purple-block" } }
<Paragraph invert>
  Call Screen requires a subscription.{' '}
  <Link href="#" invert>
    Log in
  </Link>{' '}
  to your account and make sure you are subscribed.
</Paragraph>
```

### Using with React Router Links

When using Link with React Router, you must pass in the React Router Link component as a prop. Our intention is to have developers supply the React Router Link component whether they are on version 3 of React Router, or are using the latest React Router DOM package on version 4. This way, we avoid deep dependency conflicts.

It is recommended to set up a component wrapper in your project to avoid repetition, like so:

**LinkWrapper.jsx**

```jsx noeditor static
import React from 'react'

// React Router version 3
import { Link as ReactRouterLink } from 'react-router'

// React Router version 4
// import { Link as ReactRouterLink } from 'react-router-dom'

import TdsLink from '@tds/core-link'

// If you're importing from @telus/tds, write this as:
// import { Link as TdsLink } from '@telus/tds';

const LinkWrapper = ({ children, ...rest }) => (
  <TdsLink {...rest} reactRouterLinkComponent={rest.to ? ReactRouterLink : undefined}>
    {children}
  </TdsLink>
)

export default LinkWrapper
```

**App.js**

```js noeditor static
import Link from './LinkWrapper'

const MyApp = () => (
  <main>
    ...
    <Link to="/more">Read More</Link>
    <Link href="https://www.telus.com/">Telus Website</Link>
  </main>
)
```

This way, you can output normal anchors or react router links using a single component.
