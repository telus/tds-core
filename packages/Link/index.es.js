import Link from './dist/index.es'

export default Link
