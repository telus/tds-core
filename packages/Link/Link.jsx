import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { colorGreyShark, colorWhite, colorGainsboro, colorGreyRaven } from '@tds/core-colours'
import { links } from '@tds/shared-styles'
import { safeRest } from '@tds/util-helpers'
import { componentWithName } from '@tds/util-prop-types'

import { withForwardedRef } from '@tds/shared-hocs'

import { warn } from '../../shared/utils/warn'

const base = {
  ...links.focusOutline,
  '&:link,&:visited': {
    color: colorGreyShark,
    textDecoration: 'underline',
  },
  '&:hover': {
    textDecoration: 'none',
  },
  '& svg': {},
}

const states = ({ invert }) => {
  return {
    '&:active': {
      color: invert && colorGainsboro,
      backgroundColor: invert ? 'rgba(0,0,0,0.4)' : colorGainsboro,
      borderRadius: '0.25rem',
      padding: '0.125rem',
      margin: '-0.125rem',
      textDecoration: 'underline',
    },
    '&:focus': {
      border: `0.125rem solid ${invert ? colorWhite : colorGreyRaven}`,
      padding: '0.125rem',
      margin: '-0.25rem', // (border + padding) * -1
      borderRadius: '0.25rem',
      outline: 'none',
    },
  }
}
const StyledLink = styled.a(
  base,
  {
    '& svg': {
      transition: 'transform 150ms ease-in-out',
    },
    '&:hover svg': {
      transform: 'scale(1.1, 1.1)',
    },
    '&:active svg': {
      transform: 'scale(1, 1)',
    },
  },
  ({ invert, context }) => {
    if (context.inheritColor) {
      return {
        '&:link,&:visited': {
          color: 'inherit',
        },
      }
    }
    if (invert) {
      return {
        '&:link,&:visited': {
          color: colorWhite,
        },
      }
    }
    return {}
  },
  states,
  ({ hasIcon }) => {
    if (hasIcon) {
      return {
        display: 'inline-block',
        '& > svg': {
          verticalAlign: 'bottom',
        },
      }
    }
    return {}
  }
)

/**
 * @version ./package.json
 */
const Link = (
  { reactRouterLinkComponent, invert, children, forwardedRef, icon: Icon, iconPosition, ...rest },
  context
) => {
  if (!(reactRouterLinkComponent && rest.to) && (reactRouterLinkComponent || rest.to)) {
    warn('Link', 'The props `reactRouterLinkComponent` and `to` must be used together.')
  }

  const renderChildren = useCallback(() => {
    if (Icon) {
      return (
        <>
          {iconPosition === 'left' && (
            <Icon color={invert ? 'white' : 'greyShark'} style={{ marginRight: '0.5rem' }} />
          )}
          {children}
          {iconPosition === 'right' && (
            <Icon color={invert ? 'white' : 'greyShark'} style={{ marginLeft: '0.25rem' }} />
          )}
        </>
      )
    }
    return children
  }, [children, Icon, iconPosition, invert])

  return (
    <StyledLink
      {...safeRest(rest)}
      as={reactRouterLinkComponent || 'a'}
      invert={invert}
      context={context}
      ref={forwardedRef}
      hasIcon={!!Icon}
    >
      {renderChildren()}
    </StyledLink>
  )
}
Link.propTypes = {
  /**
   * React Router Link component.
   */
  reactRouterLinkComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  /**
   * Target URL (if using 'reactRouterLinkComponent')
   */
  to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  /**
   * Target URL.
   */
  href: PropTypes.string,
  /**
   * Invert link style to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * Link text.
   */
  children: PropTypes.node.isRequired,
  /**
   * @ignore
   */
  forwardedRef: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  /**
   * Provide an icon from the Dependent icon group in `@tds/core-interactive-icon`.
   * @since 2.2.0
   */
  icon: componentWithName('Dependent', true),
  /**
   * When `icon` is provided, use `iconPosition` to place the Icon to the left or right side of Link.
   * @since 2.2.0
   */
  iconPosition: PropTypes.oneOf(['left', 'right']),
}
Link.defaultProps = {
  reactRouterLinkComponent: null,
  to: null,
  href: null,
  invert: undefined,
  forwardedRef: undefined,
  icon: undefined,
  iconPosition: 'left',
}

Link.contextTypes = {
  inheritColor: PropTypes.bool,
}

export default withForwardedRef(Link)
