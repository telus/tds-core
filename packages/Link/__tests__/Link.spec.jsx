import React from 'react'
import { shallow } from 'enzyme'

import { warn } from '../../../shared/utils/warn'

import Link from '../Link'

jest.mock('../../../shared/utils/warn')

describe('Link', () => {
  const doShallow = (overrides = {}) => shallow(<Link {...overrides}>Go home</Link>)

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders', () => {
    const link = doShallow()

    expect(link).toMatchSnapshot()
  })

  it('is an anchor HTML element when using the href attribute', () => {
    const link = doShallow({ href: 'http://telus.com' })

    expect(link).toHaveDisplayName('a')
    expect(link).toHaveProp('href', 'http://telus.com')
  })

  it('renders a react router link element when passed as a prop', () => {
    const MyLink = () => <span />
    const link = doShallow({ reactRouterLinkComponent: MyLink })

    expect(link).toMatchSelector('MyLink')
  })

  it('must use `reactRouterLinkComponent` and `to` props together', () => {
    const MyLink = () => <span />
    let link = doShallow({ reactRouterLinkComponent: MyLink })

    expect(warn).toHaveBeenCalled()

    jest.clearAllMocks()

    link = doShallow({ to: '/about' })

    expect(link).toHaveProp('to')
    expect(warn).toHaveBeenCalled()
  })

  it('can be displayed with the default styles', () => {
    const link = doShallow()

    expect(link).toHaveClassName('base')
  })

  it('can be inverted', () => {
    const link = doShallow({ invert: true })

    expect(link).toHaveClassName('inverted')
  })

  it('passes additional attributes to the link element', () => {
    const link = doShallow({ id: 'the-link', role: 'button' })

    expect(link).toHaveProp('id', 'the-link')
    expect(link).toHaveProp('role', 'button')
  })

  it('does not allow custom CSS', () => {
    const link = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(link).not.toHaveProp('className', 'my-custom-class')
    expect(link).not.toHaveProp('style')
  })
})
