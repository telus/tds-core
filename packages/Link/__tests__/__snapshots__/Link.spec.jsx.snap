// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders 1`] = `
<a
  className="base"
  href={null}
  to={null}
>
  Go home
</a>
`;
